### ISSUE TITLE
False positive `dataclass-field-order` for `dataclass_transform` wrapper with `kw_only=True`

### ISSUE THREAD

**Issue by liblaf** [astral-sh/ty#3115]
### Summary

`ty` reports a false-positive `dataclass-field-order` error for a custom `@dataclass_transform` wrapper around `dataclasses.dataclass` when the wrapper is used as `@define(kw_only=True)`.

In this case, the required field should be allowed after a field with a default, because `kw_only=True` makes the fields keyword-only. I would expect this code to produce no diagnostics.

Minimal repro:

```python
import dataclasses
from collections.abc import Callable
from typing import Any, dataclass_transform, overload


@overload
@dataclass_transform(field_specifiers=(dataclasses.field,))
def define[T: type](maybe_cls: T, /, **kwargs) -> T: ...
@overload
@dataclass_transform(field_specifiers=(dataclasses.field,))
def define[T: type](maybe_cls: None = None, /, **kwargs) -> Callable[[T], T]: ...
def define(*args, **kwargs) -> Any:
    return dataclasses.dataclass(*args, **kwargs)


@define(kw_only=True)
class A:
    foo: int = dataclasses.field(default=0)
    # ! error[dataclass-field-order]: Required field `bar` cannot be defined after fields with default values
    bar: int = dataclasses.field()
```

### Command

```bash
ty check repro.py
```

### Relevant settings

No special configuration is required to reproduce this issue.

### Playground link

<https://play.ty.dev/1bc71c1b-c147-496e-be5d-1803c86d32c4>

### Version

ty 0.0.24 (876233049 2026-03-19)

**Comment by liblaf** [comment]
I found a workaround that may help narrow down the bug.

If I explicitly declare `kw_only` in the overloads for the `@dataclass_transform` wrapper, the false-positive goes away:

```python
import dataclasses
from collections.abc import Callable
from typing import Any, dataclass_transform, overload


@overload
@dataclass_transform(field_specifiers=(dataclasses.field,))
def define[T: type](maybe_cls: T, /, *, kw_only: bool = False, **kwargs) -> T: ...
@overload
@dataclass_transform(field_specifiers=(dataclasses.field,))
def define[T: type](
    maybe_cls: None = None, /, *, kw_only: bool = False, **kwargs
) -> Callable[[T], T]: ...
def define(*args, **kwargs) -> Any:
    return dataclasses.dataclass(*args, **kwargs)


@define(kw_only=True)
class A:
    foo: int = dataclasses.field(default=0)
    # All checks passed!
    bar: int = dataclasses.field()
```

**Comment by sharkdp** [comment]
> If I explicitly declare `kw_only` in the overloads for the `@dataclass_transform` wrapper, the false-positive goes away:

Yes, indeed. ty currently uses that information to tell whether or not `kw_only` is actually supported by the dataclass-transformer in question (it basically asks: is there a named parameter called `kw_only` that we can match this argument to? if so, set `kw_only` to `True`). The examples in the [typing spec also use this pattern](https://typing.python.org/en/latest/spec/dataclasses.html#id2) of explicitly specifying e.g. `kw_only: bool = False`. Unfortunately, the typing spec is not completely clear about whether or not it is *required* for the custom decorator function to actually specify those parameters by name, or if `**kwargs` should be supported. But it does say (emphasis mine):

> kwargs allows arbitrary **additional** keyword args to be passed to dataclass_transform. This gives type checkers the freedom to support experimental parameters without needing to wait for changes in typing.py. Type checkers should report errors for any unrecognized parameters.

I will try to get this question clarified. 


Side remark: [adding `@dataclass_transform` to multiple overloads](https://peps.python.org/pep-0681/#the-dataclass-transform-decorator) may also be problematic if you aim for support with multiple type checkers:

> If the function has overloads, the dataclass_transform decorator can be applied to the implementation of the function or any one, but not more than one, of the overloads. When applied to an overload, the dataclass_transform decorator still impacts all usage of the function.

**Comment by liblaf** [comment]
@sharkdp Thanks for the side remark, I wasn't aware of that. 

I noticed that Pyright and Pyrefly don't require an explicit `kw_only` parameter in this context. However, I'm not sure if we should follow their behavior or if ty should maintain its own stricter check.
